### DISCUSSION TITLE
how to create image shadow, not text shadow

### DISCUSSION THREAD

**Discussion by zw20045500** [libvips/libvips#2385] [Q&A]
hello, how create image shadow , it likes the picture in css
![图片](https://user-images.githubusercontent.com/30209916/128604307-80587de3-fc2b-4a99-ba44-3565389c5bb3.png)

**Comment by jcupitt** [comment]
Hi @zw20045500,

Make three layers (image, shadow and background), then composite them together. The simplest way to make the shadow is to blur the alpha from the image.

What language do you need? Python?

**Reply by zw20045500** [reply]
C language, can you give me example code. thank you!

**Reply by zw20045500** [reply]
Python is also ok

**Comment by jcupitt** [comment]
I made you a python example:

```python
#!/usr/bin/python3

import sys
import pyvips

shadow_colour = [128, 0, 0]
shadow_offset = 50
shadow_radius = 10
background_colour = [255, 128, 255]

image = pyvips.Image.new_from_file(sys.argv[1], access="sequential")

# make sure the image has an alpha
if not image.hasalpha():
    image = image.addalpha() 
    
# extract the image alpha to use as the shape of the shadow
shadow_alpha = image[-1]

# enlarge and blur to make the shadow alpha
shadow_alpha = shadow_alpha.embed(shadow_radius, shadow_radius,
                                  shadow_alpha.width + 2 * shadow_radius, 
                                  shadow_alpha.height + 2 * shadow_radius)
shadow_alpha = shadow_alpha.gaussblur(shadow_radius)

# the shadow rgb is just a block of solid colour
shadow_rgb = shadow_alpha.new_from_image(shadow_colour)

shadow = shadow_rgb.bandjoin(shadow_alpha)

# the background is a block of solid colour large enough to hold everything
background = shadow.new_from_image(background_colour)
background = background.embed(0, 0, 
                              background.width + shadow_offset,
                              background.height + shadow_offset,
                              extend="copy")

# composite the three layers together
final = background.composite([shadow, image], "over", 
                             x=[shadow_offset, shadow_offset + shadow_radius],
                             y=[shadow_offset, shadow_offset + shadow_radius])

final.write_to_file(sys.argv[2])
```

Run like this:

```
$ ./drop-shadow.py ~/pics/logo4.png x.png
```

To make images like this:

![x](https://user-images.githubusercontent.com/580843/128689144-520a4df1-eebf-40b7-a303-6b987c3cc302.png)

That's rendering an RGB image. If you want an RGBA image, just remove the background stuff.

**Reply by zw20045500** [reply]
thank you your replay, but that is not my expect.  I do not konow how to remove the background corner border. I want to make image like this:
![图片](https://user-images.githubusercontent.com/30209916/128957382-13956636-8e12-4e61-b464-73c5d7e76f89.png)

**Reply by jcupitt** [reply]
Set `shadow_radius` to 0.

**Reply by zw20045500** [reply]
I try it , but it not my expect.  
![图片](https://user-images.githubusercontent.com/30209916/129357280-62366a7f-18f9-4d88-9905-db9ba03c98dc.png)

**Reply by jcupitt** [reply]
It looks like your image has no alpha. Did you include these lines?

```python
# make sure the image has an alpha
if not image.hasalpha():
    image = image.addalpha() 
```

**Comment by jcupitt** [comment]
I improved the geometry calculation:

```python
#!/usr/bin/python3

import sys
import pyvips

shadow_colour = [100, 100, 100]
shadow_offset = 10
shadow_radius = 5
background_colour = [255, 255, 255]

image = pyvips.Image.new_from_file(sys.argv[1], access="sequential")

# make sure the image has an alpha
if not image.hasalpha():
    image = image.addalpha()

# extract the image alpha to use as the shape of the shadow
shadow_alpha = image[-1]

# enlarge and blur to make the shadow alpha
shadow_alpha = shadow_alpha.embed(shadow_radius, shadow_radius,
                                  shadow_alpha.width + 2 * shadow_radius,
                                  shadow_alpha.height + 2 * shadow_radius)
shadow_alpha = shadow_alpha.gaussblur(shadow_radius / 2.0)

# the shadow rgb is just a block of solid colour
shadow_rgb = shadow_alpha.new_from_image(shadow_colour)

shadow = shadow_rgb.bandjoin(shadow_alpha)

# the background is a block of solid colour large enough to hold everything
background = shadow.new_from_image(background_colour)
extra_border = max(0, abs(shadow_offset) - shadow_radius)
background = background.embed(0, 0,
                              background.width + extra_border,
                              background.height + extra_border,
                              extend="copy")

# composite the three layers together
final_offset = max(0, shadow_offset - shadow_radius)
final = background.composite([shadow, image], "over",
                             x=[final_offset,  
                                final_offset - shadow_offset + shadow_radius],
                             y=[final_offset,  
                                final_offset - shadow_offset + shadow_radius])

final.write_to_file(sys.argv[2])
```

I see:

```
$ ./drop-shadow.py ~/pics/bambi.png x.png
```

To make:

![x](https://user-images.githubusercontent.com/580843/129400406-30268386-a0b4-48fa-a184-849296df6af4.png)

**Reply by zw20045500** [reply]
Thank you!

**Reply by jcupitt** [reply]
Hi @demisweety, that's an interesting question, but could you open a new discussion? It's confusing to have several topics in one thread.

**Reply by demisweety** [reply]
> Hi @demisweety, that's an interesting question, but could you open a new discussion? It's confusing to have several topics in one thread.

i get it

**Comment by roffe** [comment]
Took me a day to figure out how to do something similar in go-vips. here is the code for any other gophers finding themselves here

```go
func Shadow(img *vips.ImageRef, sigma, offset int) error {
	if !img.HasAlpha() {
		img.AddAlpha()
	}

	shadowAlpha, err := img.Copy()
	if err != nil {
		return err
	}
	defer shadowAlpha.Close()

	if err := shadowAlpha.ExtractBand(3, 1); err != nil {
		return err
	}

	err = shadowAlpha.Embed(
		offset,
		offset,
		shadowAlpha.Width()+offset+sigma,
		shadowAlpha.Height()+offset+sigma,
		vips.ExtendBackground,
	)
	if err != nil {
		return err
	}
	if err := shadowAlpha.GaussianBlur(float64(sigma / 2)); err != nil {
		return err
	}

	shadow, err := vips.Black(shadowAlpha.Width(), shadowAlpha.Height())
	if err != nil {
		return err
	}
	defer shadow.Close()

	if err := shadow.ToColorSpace(vips.InterpretationSRGB); err != nil {
		return err
	}

	err = shadow.DrawRect(vips.ColorRGBA{R: 0x2e, G: 0x2e, B: 0x30}, 0, 0, shadowAlpha.Width(), shadowAlpha.Height(), true)
	if err != nil {
		return err
	}

	if err := shadow.BandJoin(shadowAlpha); err != nil {
		return err
	}

	err = img.Embed(0, 0,
		shadow.Width(),
		shadow.Height(),
		vips.ExtendBackground,
	)
	if err != nil {
		return err
	}

	return img.Composite(shadow, vips.BlendModeDestOver, offset, offset)
}
```
![1](https://user-images.githubusercontent.com/5434736/196791737-8d475fdf-04b5-48a3-8207-f865351fc3b4.jpg)
![2](https://user-images.githubusercontent.com/5434736/196791743-3d0e7304-df9f-4505-b52b-67bbb294b711.jpg)

**Reply by jcupitt** [reply]
Nice! But I wouldn't use `DrawRect`, it's very slow and needs a lot of memory. The draw operations are only for paintbox style applications: they don't thread, they eat memory, they break pipelines, and they block caching.

Instead, make a 1x1 pixel image of the correct colour and type, then use `Embed` in copy mode to expand it to the size you need. This is what pyvips does:

https://github.com/libvips/pyvips/blob/master/pyvips/vimage.py#L705-L732

You should find it's noticeably quicker.

**Reply by roffe** [reply]
thank you for the tip I will implement it straight away!

**Reply by roffe** [reply]
I'm sorry but I can't seem to figure out any good way to create a single pixel image in Go without doing DrawRect and paint that one pixel on it then do embed with copy. or maybe that is the way?

In all the python examples I can see how arbitrary bands are setup from slices and joined together. But I am yet to figure out how to do this with the go-vips package

how would something like this compare perf wise to vips.Black -> ToColorSpace -> DrawRect(1,1) -> Embed(copy) -> BandJoin with Alpha

```go
...
	shadow, err := vips.Black(1, 1)
	if err != nil {
		return err
	}
	defer shadow.Close()

	if err := shadow.ToColorSpace(vips.InterpretationSRGB); err != nil {
		return err
	}

	if err := shadow.EmbedBackground(0, 0, shadow.Width(), shadow.Height(), &vips.Color{R: 0x2e, G: 0x2e, B: 0x30}); err != nil {
		return err
	}

	if err := shadow.BandJoin(shadowAlpha); err != nil {
		return err
	}
...
```

**Reply by jcupitt** [reply]
Maybe the php-vips equivalent is easier to translate:

```php
    public function newFromImage($value): Image
    {
        $pixel = static::black(1, 1)->add($value)->cast($this->format);
        $image = $pixel->embed(
            0,
            0,
            $this->width,
            $this->height,
            ['extend' => Extend::COPY]
        );
        return $image->copy([
            'interpretation' => $this->interpretation,
            'xres' => $this->xres,
            'yres' => $this->yres,
            'xoffset' => $this->xoffset,
            'yoffset' => $this->yoffset
        ]);
    }
```

And `add()` is defined as:

```php
    public function add($other, array $options = []): Image
    {
        if (self::isImageish($other)) {
            return VipsOperation::call('add', $this, [$other], $options);
        } else {
            return $this->linear(1, $other, $options);
        }
    }
```

So `$image->add([1, 2, 3])` becomes `$image->linear(1, [1, 2, 3])`, and `linear(number, array)` becomes `linear([number] * array.length, array)`, ie. `linear([1, 1, 1], [1, 2, 3])`. When libvips sees a one band image and an n-element argument to linear, it'll automatically expand the image to the same number of bands as the array of constants.

You don't need to call `colourspace` -- libvips images are just big uninterpreted arrays of values, with a tag on saying how the values should be interpreted. As long as you have the numbers in the right range (eg. for RGB you need 0 - 255, three bands), you can just set the interpretation tag with copy, which is very quick.

So, I'd do Black -> Add -> Cast -> Embed(image.size, copy) -> Copy(image.metadata) to make the constant image, then Bandjoin the alpha to that.

`image.NewFromImage([1, 2, 3])` would probably be a good thing to add to go-vips -- making constant images quickly is very handy. Python (for example) lets you do stuff like `(image > 12).ifthenelse(14, [7, 6, 8])` and you need a fast image builder for that.

**Reply by roffe** [reply]
Thank you very much for the elaborate explanation!

 I Think the problem is how go-vips has implemented `Add()` 

```
func (*vips.ImageRef).Add(addend *vips.ImageRef) error
Add calculates a sum of the image + addend and stores it back in the image
```

it only takes annoter *vips.ImageRef and I only seem to be able to create these from actual files and not an arbitrary slice of values. I'll dig some more and hopefully figure this out now that my Vips understanding is broadening :)

And yes, a NewFromImage like you are showing would be golden addition to go-vips, I've raised a question in the gitrepo for it now

**Reply by jcupitt** [reply]
Ah, Go doesn't have method overloads, so I think you'd need to call `Linear` instead.
